tests/src/providers/testqgsvirtualrasterprovider.cpp

```diff
@@ -0,0 +1,232 @@
+
+  QCOMPARE( decodedParams.width, 373 );
+  QCOMPARE( decodedParams.height, 350 );
+  QCOMPARE( decodedParams.extent, QgsRectangle( 18.6662979442000001, 45.7767014376000034, 18.7035979441999984, 45.8117014376000000 ) );
+  QCOMPARE( decodedParams.crs, QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ) );
+  QCOMPARE( decodedParams.formula, QStringLiteral( "\"dem@1\" + 200" ) );
+
+  QCOMPARE( decodedParams.rInputLayers.at( 1 ).provider, QStringLiteral( "gdal" ) );
+  QCOMPARE( decodedParams.rInputLayers.at( 0 ).provider, QStringLiteral( "gdal" ) );
+
+  qDebug() << QStringLiteral( "Raster layer: name, uri, provider" );
+  for ( int i = 0; i < decodedParams.rInputLayers.size() ; ++i )
+  {
+    qDebug() << decodedParams.rInputLayers.at( i ).name << ", " <<
+             decodedParams.rInputLayers.at( i ).uri  << ", " <<
+             decodedParams.rInputLayers.at( i ).provider;
+  }
+
+}
+
+void TestQgsVirtualRasterProvider::testUriEncoding()
+{
+
+  QgsRasterDataProvider::VirtualRasterParameters decodedParams = QgsVirtualRasterProvider::decodeVirtualRasterProviderUri( QStringLiteral( "?crs=EPSG:4326&extent=18.6662979442000001,45.7767014376000034,18.7035979441999984,45.8117014376000000&width=373&height=350&formula=\"dem@1\" + 200&dem:uri=/home/franc/dev/cpp/QGIS/tests/testdata/raster/dem.tif&dem:provider=gdal&rband:uri=/home/franc/dev/cpp/QGIS/tests/testdata/raster/band1_byte_ct_epsg4326.tif&rband:provider=gdal" ) );
+  qDebug() << QgsVirtualRasterProvider::encodeVirtualRasterProviderUri( decodedParams );
+
+}
+
+void TestQgsVirtualRasterProvider::testConstructorWrong()
+{
+  //Giving an invalid uri, with more raster referencies compared to the raster.ref that are present in the formula
+  QString str1 = QStringLiteral( "?crs=EPSG:4326&extent=18.6662979442000001,45.7767014376000034,18.7035979441999984,45.8117014376000000&width=373&height=350&formula=\"dem@1\" + 200&dem:provider=gdal&landsat:provider=gdal" );
+  QString uri = QString( "%1&%2&%3" ).arg( str1, QStringLiteral( "dem:uri=" ) % mTestDataDir % QStringLiteral( "raster/dem.tif" ),
+                QStringLiteral( "landsat:uri=" ) % mTestDataDir % QStringLiteral( "landsat.tif" ) );
+  std::unique_ptr< QgsRasterLayer > layer = std::make_unique< QgsRasterLayer >( uri,
+      QStringLiteral( "layer" ),
+      QStringLiteral( "virtualraster" ) );
+
+  if ( layer->dataProvider()->isValid() )
+  {
+    QVERIFY( layer->dataProvider()->isValid() );
+    QVERIFY( layer->isValid() );
+
+  }
+  else
+  {
+    QVERIFY( ! layer->dataProvider()->isValid() );
+    QVERIFY( ! layer->isValid() );
+    qDebug() << QStringLiteral( "The dataprovider is not valid" );
+  }
```

Comment:
with this if/else approach I think we are not actually auto-testing things - let's replace it by just something like this:
```c++
    QVERIFY( !layer->isValid() );
```